bin/utest/vm_map.c

```diff
@@ -0,0 +1,79 @@
+#include "utest.h"
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+int test_sharing_memory_simple(void) {
+  char *map =
+    mmap(NULL, 4096, PROT_READ | PROT_WRITE, MAP_ANON | MAP_SHARED, 0, 0);
+
+  assert(map != (char *)MAP_FAILED);
+
+  pid_t pid = fork();
+
+  assert(pid >= 0);
+
+  if (pid == 0) {
+    /* child */
+    strcpy(map, "Hello, World!");
+    exit(0);
+  } else {
+    /* parent */
+    int status;
+    waitpid(-1, &status, 0);
+  }
+
+  assert(strcmp(map, "Hello, World!") == 0);
+
+  assert(munmap(map, 4096) == 0);
+  return 0;
+}
+
+int test_sharing_memory_child_and_grandchild(void) {
+  char *map =
+    mmap(NULL, 4096, PROT_READ | PROT_WRITE, MAP_ANON | MAP_SHARED, 0, 0);
+
+  assert(map != (char *)MAP_FAILED);
+
+  pid_t pid = fork();
+
+  assert(pid >= 0);
+
+  if (pid == 0) {
+    /* child */
+    pid = fork();
+
+    assert(pid >= 0);
+
+    if (pid == 0) {
+      /* grandchild */
+      strcpy(map, "Hello from grandchild!");
+      exit(0);
+    } else {
+      /* child */
+      int status;
+      waitpid(-1, &status, 0);
+      assert(WIFEXITED(status));
+      assert(strcmp(map, "Hello from grandchild!") == 0);
```

Comment:
Probably `WIFEXITED` doesn't check anything on `waitpid` failure.

Reply:
Maybe you're right. I added assert to check if waitpid succeeded.